### DISCUSSION TITLE
Reliable hang of SQLAlchemy

### DISCUSSION THREAD

**Discussion by dwt** [sqlalchemy/sqlalchemy#7848] [Usage Questions] [answered]
### Describe the bug

Not quite sure if this is even a realistic problem.

What did I do? I played around with sqlalchemy while stopping the database server mid connection. To understand how SQLAlchemy behaves.

Doing that I discovered that I could easily hang SQLAlchemy for a very long time (I gave up after several minutes).

### To Reproduce

https://gist.github.com/dwt/0de420d56ba6727462b66fbe02606609

### Error

```
SQLAlchemy hangs
``` 

### Versions

- OS: debian buster
- Python: 3.6
- SQLAlchemy: 1.4.32
- Database: postgres
- DBAPI psycopg2

### Additional context

I've built the reproduction as a docker container as that is how I initially encountered the problem.

**Comment by dwt** [comment]
If you need any additional details, feel free to ask. I'm honestly a bit lost if that is even a problem or just an artefact of the way I played around with sqlalchemy here.

**Comment by zzzeek** [comment]
there's nothing in SQLAlchemy that's capable of "hanging", so what it looks like here is that via some combination of killing the PostgreSQL server and working with interprocess communication (all of which are complete minefields for things hanging), you've gotten either your psycopg2 connection or your Python process to hang.    Try pressing ctrl-C, assuming that allows the program to break out, and it will generate a stack trace showing where the program is stuck.  remove the call to create_engine() and try using the psycopg2 API directly also.

**Reply by dwt** [reply]
Well, I've had completely no luck with ctrl-c or ctrl-z, none of these where able to kill the process. `kill -9` was the only one that allowed me to get out of this.

Using psycopg2 directly might be a good idea, I can take a look at that later.

Just to understand what should be happening: Should sqlalchemy be resilient against these kinds of hangs? Is that even a category of bug you are working on?

**Reply by zzzeek** [reply, marked as answer]
> Well, I've had completely no luck with ctrl-c or ctrl-z, none of these where able to kill the process. `kill -9` was the only one that allowed me to get out of this.

take a look at your PostgreSQL server process list, or query from [pg_stat_activity](https://www.postgresql.org/docs/14/monitoring-stats.html) to see what locks / transactions / statements may be idle in transaction.


> Just to understand what should be happening: Should sqlalchemy be resilient against these kinds of hangs? Is that even a category of bug you are working on?

SQLAlchemy is a library that sends commands to a Python DBAPI like psycopg2.  There's no capability for a DBAPI wrapper to be able to detect and anticipate application deadlocks, any more than the driver itself can.  Even the Python interpreter itself can't do this, if you write a program that calls a threading.Lock() in a reentrant fashion, your program hangs.   

The code being illustrated here has *many* things happening that are widely known to be things that can cause deadlocks and hangs.  It's very complex. I'm not sure what the purpose of all this code is but if this is the kind of thing you are working on, there are debugging skills and techniques that would be useful to make use of; try out a module like [faulthandler](https://docs.python.org/3/library/faulthandler.html) so that you can send a signal to the app to dump a traceback, for example.

**Reply by dwt** [reply]
Thanks a lot for the feedback, this made me understand that what I'm trying to test is likely a bad idea.

**Reply by johanvanderkuijl** [reply]
I don't think this is a bad idea, since I encounter the same issue in production. It might be due to the configuration of our firewall. I can reproduce the issue by starting my application in docker-compose, and then _pause_ the postgres container. The python client hangs then, often exactly 260 seconds on my dev environment, and 16 minutes on the production environment.
After unpausing the postgres container, the client immediately resumes.
